src/node/transaction.cpp

```diff
@@ -36,10 +36,19 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
         if (!existingCoin.IsSpent()) return TransactionError::ALREADY_IN_CHAIN;
     }
     if (!node.mempool->exists(hashTx)) {
-        // Transaction is not already in the mempool. Submit it.
+        // Transaction is not already in the mempool.
         TxValidationState state;
+        CAmount fee;
+        bool test_accepted = AcceptToMemoryPool(*node.mempool, state, std::move(tx),
```

Comment:
Both this ATMP call and the one below should have their `std::move(tx)` argument changed to `tx`. In general, you should never use a variable after it's been used with `std::move` in a function call, since it might have been moved away. In this case, if `tx` has been moved from, it'll be null, and any attempt to dereference it will segfault.

However, because ATMP takes a _const_ reference to `CTransactionRef`, it can't be moved away from, so if ATMP wants to retain shared ownership of the transaction, it must copy rather than move to the new shared_ptr. That means that technically it is safe to use `tx` after this call.

Let's just make it explicit though: ATMP takes a const ref and has to copy if it wants to retain shared ownership, so just pass it `tx`.

See R30, R34, R35 and R36 here: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r30-take-smart-pointers-as-parameters-only-to-explicitly-express-lifetime-semantics for more background on passing raw/smart pointers.

Reply:
Thanks for noting this and posting the link, @jnewbery ! After looking into this, I have two follow-up questions for you if that's OK:

#### Question 1 (historical context)

Previously, this call always used `std::move(tx)` because `tx` did not appear later in the function. However, use of `tx` after `move` appeared in c7eb6b4f1fe5bd76388a023529977674534334a7, which was merged as part of #18044. Do you happen to have any ideas about why it wasn't a concern then?

#### Question 2 (C++ guidelines)

The linked guidelines mention the `const shared_ptr<widget>&` case, but not the `const shared_ptr<widget>` case that's used in `BroadcastTransaction`:

<img width="813" alt="guidelines" src="https://user-images.githubusercontent.com/4276679/91061181-b88e7400-e5e0-11ea-8c86-1e50908c689e.png">

Do you know if we should think of the `const shared_ptr<widget>` case any differently?

Reply:
> However, use of tx after move appeared in c7eb6b4, which was merged as part of #18044. Do you happen to have any ideas about why it wasn't a concern then?

Because it was missed in review. In this case, it's not going to cause any issues because `tx` isn't actually moved from, but we should still never be doing anything with a variable after it's had `std::move` used on it.

> Do you know if we should think of the const shared_ptr<widget> case any differently?

I don't think there's any good reason to pass a shared_ptr by const value. I may be mistaken here, but I think that means you need to increment the refcount when you enter the function (because the parameter is copy-constructed from the argument), and if you want to retain the refcount, you need to increment it again (because you can't move from the const parameter). Move semantics with smart pointers is a deep rabbit hole. There's lots more here: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/ and http://scottmeyers.blogspot.com/2014/07/should-move-only-types-ever-be-passed.html.

Reply:
done! thanks for the explanation @jnewbery :)